File: backend/app/models/acl_request.py
from sqlalchemy import Column, Integer, String, DateTime, Enum as SQLEnum
from sqlalchemy.sql import func
import enum
from app.db.base import Base

class RequestStatus(str, enum.Enum):
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"

class AclRequest(Base):
    __tablename__ = "acl_requests"

    id = Column(Integer, primary_key=True, index=True)
    username = Column(String, index=True, nullable=False)
    operation = Column(String, nullable=False)  # GET, POST, DELETE (API Operation)
    kafka_operation = Column(String, nullable=True) # READ, WRITE, ALL, etc. (Kafka Operation)
    permission_type = Column(String, nullable=True) # ALLOW, DENY
    pattern_type = Column(String, nullable=True, default="LITERAL") # LITERAL, PREFIXED
    resource_type = Column(String, nullable=False) # TOPIC, etc.
    resource_name = Column(String, nullable=False)
    principal = Column(String, nullable=True) # User:alice
    host = Column(String, nullable=True, default="*")
    cluster = Column(String, nullable=False, default="default")
    status = Column(SQLEnum(RequestStatus), default=RequestStatus.PENDING, nullable=False)
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    approved_by = Column(String, nullable=True) # Admin username
    admin_comment = Column(String, nullable=True) # Admin comment on approve/reject
    @property
    def request_type(self):
        if self.operation == "POST":
            return "Create ACL"
        elif self.operation == "DELETE":
            return "Delete ACL"
        return self.operation

File: KafkaManagerUI_SpringBoot/src/main/java/com/kafkaui/entity/AclRequest.java
package com.kafkaui.entity;

import jakarta.persistence.*;
import lombok.Data;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;

@Entity
@Table(name = "acl_requests")
@Data
public class AclRequest {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(nullable = false)
    private String username;

    @Column(nullable = false)
    private String operation;

    @Column(name = "kafka_operation")
    private String kafkaOperation;

    @Column(name = "permission_type")
    private String permissionType;

    @Column(name = "pattern_type")
    private String patternType = "LITERAL";

    @Column(name = "resource_type", nullable = false)
    private String resourceType;

    @Column(name = "resource_name", nullable = false)
    private String resourceName;

    private String principal;
    private String host = "*";

    @Column(nullable = false)
    private String cluster = "default";

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name = "status", nullable = false)
    private RequestStatus status = RequestStatus.PENDING;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "approved_by")
    private String approvedBy;

    @Column(name = "admin_comment")
    private String adminComment;
}
